types/rdfjs__term-map/index.d.ts

```diff
@@ -0,0 +1,16 @@
+// Type definitions for @rdfjs/term-map 1.0
+// Project: https://github.com/rdfjs-base/term-map
+// Definitions by: tpluscode <https://github.com/tpluscode>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import { Term } from "rdf-js";
+
+interface TermMap<T extends Term = Term, V = any> extends Map<T, V> {
+}
+
+// tslint:disable-next-line no-unnecessary-class
+declare class TermMap<T extends Term = Term, V = any> {
+    constructor(entries?: Array<[Term, V]>);
+}
```

Comment:
```suggestion
declare class TermMap<T extends Term = Term, V = any> extends Map<T, V> {
    constructor(entries?: Array<[Term, V]>);
}
```

Reply:
By declaring the interface and class under the same name I do not have to repeat all the members in the class body.
I thought this was the "whole point of declaration merging" 🙄

Reply:
Extending `Map<T, V>` is what lets you avoid repeating members (note I moved the heritage clause to the class declaration in my suggestion). Declaration merging lets you treat multiple declarations as a single symbol (which in fact is exactly what happens under the hood). Sorry I didn’t look far enough ahead to realize in my earlier review that the interface was going to be useless. When I said

> There’s no reason to distinguish between the interface and the class—that’s the whole point of declaration merging.

what I meant was that it is _not possible_ to “distinguish” between multiple merged declarations, because each declaration simply contributes to a shared symbol. What you reference is that shared symbol, and the symbol takes on the meanings and members of all its constituent declarations.

To illustrate, here’s a simplified representation of how the merging works for the code as-is:

```ts
// Meanings: Type (i.e., can be used in a type position, not in a value or namespace position)
// Members: { ...inherit from Map<T, V> }
interface TermMap<T extends Term = Term, V = any> extends Map<T, V> {
}

// Meanings: Type | Value (can be used as a type, but also as a value, as in `new TermMap()`)
// Members: {}
declare class TermMap<T extends Term = Term, V = any> {
    constructor(entries?: Array<[Term, V]>);
}

/* MERGE RESULT */
// Meanings: Type | Value
// Members: { ...inherit from Map<T, V> }
```

But you can notice that this distribution of meanings and members has some redundancy. The class contributes no _members_, and the interface contributes no _meanings_ that the class didn’t already have. The interface cannot possibly take on a value meaning (much less one with a constructor), but the class can certainly take on the members that the interface contributes. By moving the `extends Map<T, V>` from the interface declaration to the class declaration, the class declaration is now

```ts
// Meanings: Type | Value
// Members: { ...inherit from Map<T, V> }
declare class TermMap<T extends Term = Term, V = any> extends Map<T, V> {
    constructor(entries?: Array<[Term, V]>);
}
```

which, in a single declaration, has the exact same meanings and members as the merged symbol from before. So, there’s no reason to split them up into two declarations.

Again, sorry I didn’t look two steps ahead to avoid this intermediate revision. Hope this clears things up.